src/Server.cpp

```diff
@@ -313,73 +565,108 @@ void Server::acceptConnections(const Socket& serverSock, uint32_t eventMask)
+void handleConnection(Server& server, const int clientFd, Connection& connection)
+{
+	LOG_DEBUG << "Handling connection: " << connection.m_clientSocket << " for server: " << connection.m_serverSocket;
+
+	switch (connection.m_status) {
+	case (Connection::ReceiveRequest):
+		connectionReceiveRequest(server, clientFd, connection);
+		break;
+	case (Connection::ReceiveBody):
+		connectionReceiveBody(server, clientFd, connection);
+		break;
+	case (Connection::BuildResponse):
+		connectionBuildResponse(server, clientFd, connection);
+		break;
+	case (Connection::SendResponse):
+		connectionSendResponse(server, clientFd, connection);
+		break;
+	case (Connection::Timeout):
+		connectionHandleTimeout(server, clientFd, connection);
+		break;
+	case (Connection::Closed):
+		break;
+	}
+}
+
+/**
+ * @brief Receive a request from a client.
+ *
+ * This function reads data from the client socket using Server::readFromSocket().
+ * The amount of bytes to read from the socket is determined by the bytes already received from the client. It can never
+ * be bigger than the buffer size.
+ * - If bytes read is -1, it indicates an internal server error
+ * - If bytes read is 0, it indicates that the connection has been closed by the client
+ * In both cases the clientFd is closed and the connection status is set to Closed.
+ * If bytes read is greater than 0, the bytes received are added to the connection buffer and the bytes received of the
+ * connection are updated.
+ * If the buffer contains a complete request, it parses the request. Then the request is removed from the client buffer,
+ * the connection is set to BuildResponse state and the event is modified to listen to EPOLLOUT.
+ * If no complete request was received and the received bytes match the buffer size, sets HTTP status code to 413
+ * Request Header Fields Too Large and status to build response, since the request header was too big.
+ *
+ * In any case the time since the last event is updated to the current time.
+ *
+ * @param server The server object which handles the connection.
+ * @param clientFd The file descriptor of the client.
+ * @param connection The connection object to receive the request for.
+ * @todo Implement body handling.
+ * @todo Send error message if request is too big.
+ */
+void connectionReceiveRequest(Server& server, int clientFd, Connection& connection)
 {
-	LOG_DEBUG << "Handling connection: " << connection.getClient() << " for server: " << connection.getServer();
-	// Handle client data
-	char buffer[s_bufferSize];
-	HTTPRequest request;
-
-	request.method = MethodCount;
-	request.httpStatus = StatusOK;
-	request.shallCloseConnection = false;
-
-	const ssize_t bytesRead = recv(connection.getClient().fd, buffer, s_bufferSize, 0);
-	if (bytesRead < 0) {
-		std::cerr << "error: read\n";
-		close(connection.getClient().fd);
+	LOG_DEBUG << "ReceiveRequest for: " << connection.m_clientSocket;
+
+	const size_t bufferSize = 1000;
```

Comment:
I think we would move complexity: we would need to make sure, that the buffer in Connection is big enough (aka std::string needs to be .resize(), to hold the data). Right now this is handled automatically since we use += to realloc the std::string.

Alternative: call .resize() before reading buffer:
- For the request headers this can be done in the constructor of Connection. Or maybe in the function after its clear which server is serving (if we want to make it variable depending on config)
- For the body it would need to be done in the function reading in body (depending on the server config)

Note: since current buffer lies on the stack, it may be faster to read into. But this may be negligible and the gainz could be destroyed by the realloc of std::string m_buffer.

I don't care. Should we read directly into Connection buffer?

Reply:
We can leave the local buffer as is as it is just used as a temp container to read from the socket and is then added to the "permanent" buffer. That part is clear to me now.

What confused me is the line

> const size_t bytesToRead = bufferSize - connection.m_bytesReceived;

If the amount of received bytes is > 1000 we wouldn't read anymore. That would impede our ability to read larger requests. There was a server directive which restricts the general request size? Not only the body, but the whole request. Can't find it rn, or we just use the default max size limit of 1 MB for the body - therefore the request cannot be larger than this. Or set it to the client_body_buffer_size if given. Then this would equal a check which returns an error if limit is reached.

Reply:
Ah! This function only reads the request **header** 

The 1k mimics client_header_buffer_size of nginx: [LINK](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_header_buffer_size)

If it detects a complete request header, determined by isCompleteRequestHeader(), then it decides if it reads body if there is one (well not right now, since not implemented yet) or if it sets status to build response.

Right now this is a rather simple implementation:
- body handling is not implemented yet > connectionReceiveBody() as well is not implemented
- client_header_buffer_size is not configurable, but fixed at 1k
- nginx also has a directive for [large_client_header_buffers ](https://nginx.org/en/docs/http/ngx_http_core_module.html#large_client_header_buffers) which activate when the request headers are too big
- nginx divides request parsing further: it checks after request line and after parsing the `host` header field, if it can select an appropriate virtual server: [LINK](https://nginx.org/en/docs/http/server_names.html#virtual_server_selection). 
This could influence the correct size of the buffers, depending on configfile

I can rename to **connectionReceiveRequestHeader**() to make it clearer.
I would leave implementation of body receiving open (since it fits thematically to #44)
I would leave the other stuff and maybe open issues for them?

Reply:
Implementation of the body handling can be effected in #44. I would create a static const int s_clientHeaderBufferSize in Server and set it to 1k plus refer in the documentation to the nginx value to make it clear that this is not an arbitrary number. And add a check for if the s_clientHeaderBufferSize is reached, in which case an error is returned to the client.

Reply:
- renamed and made static to s_clientHeaderBufferSize
- added @todo to make clientHeaderBufferSize configurable > will also create issue regarding configuration re clientHeaderBufferSize, doesn't seem like a big change
- check was already present (now with renamed buffer size):
```
if (connection.m_bytesReceived == Server::s_clientHeaderBufferSize) {
  LOG_ERROR << "Buffer full, didn't receive complete request header from " << connection.m_clientSocket;
  connection.m_request.httpStatus = StatusRequestHeaderFieldsTooLarge;
  connection.m_status = Connection::BuildResponse;
}
```